DSpace/dspace-angular
# Fix match theme by handle with canonical prefix https://hdl.handle.net/ not working

## References
* Fixes #2348
* Backend PR DSpace/DSpace#9162

## Description
This PR fixes a problem with [this regex](https://github.com/DSpace/dspace-angular/blob/b55e4327a88c993535219f6c6793a6888a015c8f/src/app/shared/handle.service.ts#L4) that assumes all handle URLs are prefixed by a `handle/`. This is not the case if you set the `handle.canonical.prefix` to https://hdl.handle.net/. The solution is to retrieve the `handle.canonical.prefix` from the backend and use it inside the regex. By default the `handle.canonical.prefix` is not exposed, so you also need the backend PR in order to test this fix.

## Instructions for Reviewers
List of changes in this PR:
* The `HandleService` now retrieves the `handle.canonical.prefix` from the backend. This means that `normalizeHandle` doesn't return a `string | null` anymore but an `Observable<string | null>`. This means that the code using this method also needs to be refactored:
  * The `CurationFormComponent` used this method, and it was fairly simple to switch to `Observable`s.
  * The `HandleTheme` also uses `normalizeHandle` in its `matches` method. So this means that the return type of the matches method needs to be changed for all the classes extending `Theme` from `boolean` ➜ `Observable<boolean>`, but this also means that more methods needed to be refactored on the two places where this method is used:
    * In the `updateThemeOnRouteChange$` method, the code that finds which theme should be used when you're not on a Community/Collection/Item page now still has to iterate over the list of available themes and return the first theme where its `matches` method returns `true`. The only difference is that it now returns an `Observable<true>` instead, so I had to wrap everything in a `concatMap` to ensure that even if the second theme returns true faster than the first theme, the first matching theme is still returned. Luckily this part of the code was already in a `switchMap`, so this refactor cascade stops here. 
    * The `matchThemeToDSOs` also uses `Theme.matches()` and had to be refactored in the same way as above. But it not only had to iterate over all the themes but for each theme, it also had to iterate over the DSO you are currently on, but also the parent DSOs. For those DSO-theme combinations, it had to use `Theme.matches()` and now also had to always return the first matching theme no matter the order in which the `Observables` resolve. This method was only used in the `updateThemeOnRouteChange$` method inside the `switchMap`, so the refactor cascade stops here. 
* I also noticed that an error is sometimes thrown in `CommunityPageSubCommunityListComponent` & `CommunityPageSubCollectionListComponent` because the destroy is sometimes called before the component has time to fully initialize. This can happen when the component initializes itself in the default theme, but then while initializing, a match is found for a `UUIDTheme`/`HandleTheme`/`RegexTheme` so the component is destroyed before the `config` property could be initialized.

Guidance for how to test/review this PR:
* To test this, we need 2 themes. You can use the 2 existing default themes for this: `custom` & `dspace` 
  * Uncomment the `CustomEagerThemeModule` in `src/themes/eager-themes.module.ts`
  * Make a change to a part of the page that is used on every page like the `HeaderComponent` (this will make testing easier). By default you can see that the `dspace` theme uses a green bottom border, while the `custom` theme uses the base theme's header by default without that green border.
* Set the `handle.canonical.prefix = https://hdl.handle.net/` and ensure that `rest.properties.exposed = handle.canonical.prefix` is also present in your config.
* Make a new community/collection/item after rebuilding your backend. They should now have https://hdl.handle.net/ in their handle URL.
* Add that handle to the custom theme in your `config.yml` like this:
  ```yml
  themes:
  - name: custom
    handle: 123456789/8
  - name: dspace
  ```
* Verify that it now works correctly when you go to https://hdl.handle.net/123456789/8 for example, but check that the theme is also applied to all the child DSOs (meaning that if you put the handle of the community in your `config.yml`, the same theme should also be applied to all its sub-communities, collections & items).
* Verify that other theme config like headTags also still work like expected because a lot of changes were made.

## Checklist
